src/future/mod.rs

```diff
@@ -11,28 +11,44 @@ use crate::runtime::thread;
 use std::future::Future;
 use std::pin::Pin;
 use std::result::Result;
-use std::task::{Context, Poll};
+use std::sync::Arc;
+use std::task::{Context, Poll, Waker};
 
 /// Spawn a new async task that the executor will run to completion.
 pub fn spawn<T, F>(fut: F) -> JoinHandle<T>
 where
     F: Future<Output = T> + Send + 'static,
     T: Send + 'static,
 {
-    let result = std::sync::Arc::new(std::sync::Mutex::new(None));
     let stack_size = ExecutionState::with(|s| s.config.stack_size);
-    let task_id = ExecutionState::spawn_future(Wrapper::new(fut, std::sync::Arc::clone(&result)), stack_size, None);
+    let inner = Arc::new(std::sync::Mutex::new(JoinHandleInner::default()));
+    let task_id = ExecutionState::spawn_future(Wrapper::new(fut, inner.clone()), stack_size, None);
 
     thread::switch();
 
-    JoinHandle { task_id, result }
+    JoinHandle { task_id, inner }
 }
 
 /// An owned permission to join on an async task (await its termination).
 #[derive(Debug)]
 pub struct JoinHandle<T> {
     task_id: TaskId,
-    result: std::sync::Arc<std::sync::Mutex<Option<Result<T, JoinError>>>>,
+    inner: std::sync::Arc<std::sync::Mutex<JoinHandleInner<T>>>,
+}
+
+#[derive(Debug)]
+struct JoinHandleInner<T> {
+    result: Option<Result<T, JoinError>>,
+    waker: Option<Waker>,
+}
+
+impl<T> Default for JoinHandleInner<T> {
+    fn default() -> Self {
+        JoinHandleInner {
+            result: None,
+            waker: None,
+        }
+    }
```

Comment:
nit: this can just be derived

Reply:
Oh right, this is the stupid thing where the derive macro always emits a `T: Default` bound even if it's not required.